### ISSUE TITLE
Add `apm deprecate` command

### ISSUE THREAD

**Issue by lee-dohm** [atom/apm#351]
The issue has come up a couple times where a package's functionality is subsumed by Atom, the author doesn't want to support a package anymore or other circumstances where a range of versions for a package no longer function, should be uninstalled (security, horrible bug discovered), or what have you.

The idea is to create an `apm deprecate` command that will:

1. Allow a package author to specify a version or version range that should be considered "deprecated" for whatever reason
1. Allow a package author to specify a deprecation message to be displayed to the user
1. Allow a package author to retrieve the cumulative version range that is deprecated for their package, if any
1. Provide warnings to the package author when common undesirable states would be created by the current command

## Command-Line Interface

```
apm deprecate [<package_name>]
apm deprecate [<package_name>] <version_range> [<message>]

Deprecates a range of versions of the given package with an optional message.

If no package name is specified, then the `package.json` in the current working directory is used to determine the package name.

If no version range is specified, the cumulative deprecated version range is displayed. For example, if one command deprecates `< 1.0` and another deprecates `>= 1.5 < 2.0`
```

Version range semantics supplied by the node-semver package: https://github.com/npm/node-semver#ranges

## Problematic or Invalid States

TBD

**Comment by lee-dohm** [comment]
Actually, that cumulative version range notation could get ugly. Probably best to emit a newline instead of `||`.

**Comment by lee-dohm** [comment]
An example of the kind of issue that this command can fix is https://github.com/atom/settings-view/issues/492. A version of my [indentation-indicator package](https://atom.io/packages/indentation-indicator) was published that allows installation on versions of Atom that did not have the v1.0 status-bar API, which is required by the version of the package. Since there are Atom users that do not fastidiously keep their Atom up-to-date, there will be people that are running Atom in the vulnerable version range. All I, as a package author, currently can do is to wait for them to report the bug and tell them to update.

If this command existed, I could deprecate the problem package version, preventing any new installations.

**Comment by izuzak** [comment]
> If this command existed, I could deprecate the problem package version, preventing any new installations.

Could you also use `apm unpublish` to unpublish those versions which you don't want anyone to install anymore?

```
$ apm unpublish --help
Usage: apm unpublish [<package_name>]
       apm unpublish <package_name>@<package_version>

Remove a published package or package version from the atom.io registry.
```

**Comment by lee-dohm** [comment]
Ah ... didn't know that you could unpublish specific versions, thanks!

**Comment by izuzak** [comment]
I think this is what @mnquintana wanted in https://github.com/atom/apm/issues/334 as well (we talked about that more in a different channel, so I'll let him correct me if I misunderstood).

**Comment by mnquintana** [comment]
Yup, this is exactly the kind of thing I was looking for! I'll post more complete thoughts later, but for now here's how npm handles it for reference: https://docs.npmjs.com/cli/deprecate

**Comment by Glavin001** [comment]
How are developers currently showing those deprecation warning messages for Atom API 1.0? I've noticed that a lot, that packages now show messages recommending other packages.
